A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,529 @@
+  state but does contain a cached resourc (i.e., we have seen an LDS or
+  CDS resource deletion but have ignored it).
+- `received_error`: The cache entry is in RECEIVED_ERROR state, and no
+  resource is cached.
+- `received_error_but_cached`: The cache entry is in RECEIVED_ERROR
+  state, and there is a cached resource (i.e., we have ignored a data
+  error due to "fail_on_data_errors" not being present).
+
+### Summary of Behavior Changes
+
+The following table shows the old and new behavior for each case.
+
+Error | Resource Already Cached | fail_on_data_errors Server Feature | Old Behavior | New Behavior
+----- | ----------------------- | ---------------------------------- | ------------ | ------------
+xDS channel reports TRANSIENT_FAILURE | false | (any) | `OnError()`, Fail data plane RPCs | `OnResourceChanged(Status)`, Fail data plane RPCs
+xDS channel reports TRANSIENT_FAILURE | true | (any) | `OnError()`, Ignore | `OnAmbientError(Status)`, Ignore
+ADS stream failed without reading a response | false | (any) | `OnError()`, Fail data plane RPCs | `OnResourceChanged(Status)`, Fail data plane RPCs
+ADS stream failed without reading a response | true | (any) | `OnError()`, Ignore | `OnAmbientError(Status)`, Ignore
+NACK from client | false | (any) | `OnError(status)`, Fail data plane RPCs | `OnResourceChanged(Status)`, Fail data plane RPCs
+NACK from client | true | false | `OnError(status)`, Ignore | `OnAmbientError(Status)`, Ignore
+NACK from client | true | true  | `OnError(status)`, Ignore | `OnResourceChanged(Status)`, Drop existing resource and fail RPCs
+Resource timeout | false | (any) | `OnResourceDoesNotExist()`, Fail data plane RPCs | `OnResourceChanged(Status(NOT_FOUND))` (status will be UNAVAILABLE if `resource_timer_is_transient_error` server feature is present), Fail data plane RPCs
+LDS or CDS resource deletion from server | true | false | `OnResourceDoesNotExist()`, Drop existing resource and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnAmbientError(Status(NOT_FOUND))`, Ignore
+LDS or CDS resource deletion from server | true | true | `OnResourceDoesNotExist()`, Drop existing resources and fail RPCs (watcher notification skipped if "ignore_resource_deletion" server feature is present) | `OnResourceChanged(Status(NOT_FOUND))`, Drop existing resource and fail RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | false | N/A | `OnAmbientError(status)`, Ignore
+[xRFC TP3] error with status NOT_FOUND or PERMISSION_DENIED | true | true | N/A | `OnResourceChanged(status)`, Drop existing resource and fail RPCs
+[xRFC TP3] error with other status | false | (any) | N/A | `OnResourceChanged(status)`, Fail data plane RPCs
+[xRFC TP3] error with other status | true | (any) | N/A | `OnAmbientError(status)`, Ignore
```

Comment:
These are N/A because we aren't currently setting `supports-resource-error-unwrapping` (which seems to be mentioned as almost an afterthought in TP3, in the "Rationale" section, oddly)?  This document should probably contain that string somewhere, right?

Reply:
Aha, I see, sorry.  So it's fine for the management server to always send the errors back, and it's N/A here because we are not yet reading the new field.

Reply:
Yes, exactly.